### PR TITLE
Feature: upper- or lowercase hexadecimal literals

### PR DESCRIPTION
**Pull request by ArjenL**
A small feature to consistently case hexadecimal literals in your Rust code.

### CODE REVIEW

**Comment by calebcartwright** [comment]
Apologies for the delay but still haven't had a chance to go through this in detail. It seems like a reasonable feature to support so I suspect we'll be able to move ahead, could just potentially have a bit of bikeshedding to do on implementation specifics.

**Comment by ArjenL** [comment]
Thanks! Some bikeshedding is fine, I'm sure some things could be better/more idiomatic.

File: Configurations.md [hunk 1]
```diff
@@ -1047,6 +1047,13 @@ fn lorem() -> usize {
 
 See also: [`tab_spaces`](#tab_spaces).
 
+## `hex_literal_case`
+
+Change the case of the letters in hexadecimal literal values
+
+- **Default value**: `Ignore`
+- **Possible values**: `ToUpper`, `ToLower`
+- **Stable**: No
```
Comment (calebcartwright) [inline Configurations.md:1056]: Couple minor things I'd like to tweak here. Let's adjust the text to reflect the fact that it can _set_ vs. implying it will change.

Instead of `Ignore`, we should use `Preserve` to be consistent with variants in our other configuration options, and then let's drop the `To` prefix on the other variants.

```suggestion
Control the case of the letters in hexadecimal literal values

- **Default value**: `Preserve`
- **Possible values**: `Upper`, `Lower`
- **Stable**: No
```

File: src/config/mod.rs [hunk 2]
```diff
@@ -569,6 +571,7 @@ license_template_path = ""
 format_strings = false
 format_macro_matchers = false
 format_macro_bodies = true
+hex_literal_case = "Ignore"
```
Comment (calebcartwright) [inline src/config/mod.rs:574]:
```suggestion
hex_literal_case = "Preserve"
```

File: src/config/options.rs [hunk 3]
```diff
@@ -127,6 +127,17 @@ pub enum ImportGranularity {
...
 
+/// Controls how rustfmt should handle case in hexadecimal literals.
+#[config_type]
+pub enum HexLiteralCase {
+    /// Leave the literal as-is
+    Ignore,
+    /// Uppercase the literal
+    ToUpper,
+    /// Lowercase the literal
+    ToLower,
```
Comment (calebcartwright) [inline src/config/options.rs:138]:
```suggestion
    Preserve,
    /// Ensure all literals use uppercase lettering
    Upper,
    /// Ensure all literals use lowercase lettering
    Lower,
```

File: src/expr.rs [hunk 4]
```diff
@@ -1210,6 +1211,28 @@ fn rewrite_string_lit(context: &RewriteContext<'_>, span: Span, shape: Shape) ->
...
 
+fn rewrite_int_lit(context: &RewriteContext<'_>, lit: &ast::Lit) -> Option<String> {
+    let span = lit.span;
+    let symbol = lit.token.symbol.as_str();
+
+    if symbol.starts_with("0x") {
+        let hex_lit = match context.config.hex_literal_case() {
+            HexLiteralCase::Ignore => None,
+            HexLiteralCase::ToUpper => Some(symbol[2..].to_ascii_uppercase()),
+            HexLiteralCase::ToLower => Some(symbol[2..].to_ascii_lowercase()),
```
Comment (calebcartwright) [inline src/expr.rs:1222]:
```suggestion
            HexLiteralCase::Preserve => None,
            HexLiteralCase::Upper => Some(symbol[2..].to_ascii_uppercase()),
            HexLiteralCase::Lower => Some(symbol[2..].to_ascii_lowercase()),
```

File: src/expr.rs [hunk 5]
```diff
@@ -1210,6 +1211,28 @@ fn rewrite_string_lit(context: &RewriteContext<'_>, span: Span, shape: Shape) ->
...
+        if let Some(hex_lit) = hex_lit {
+            return Some(format!(
+                "0x{}{}",
+                hex_lit,
+                lit.token.suffix.map_or(String::new(), |s| s.to_string())
+            ));
+        }
+    }
+
+    Some(context.snippet(span).to_owned())
```
Comment (calebcartwright) [inline src/expr.rs:1233]: Could you make sure the values will continue to be passed through `wrap_str` as they are currently? If we drop that then we're potentially introducing some breaking formatting changes around some edge cases approaching width limits

File: tests/source/hex_literal_lower.rs [hunk 6]
```diff
@@ -0,0 +1,5 @@
+// rustfmt-hex_literal_case: ToLower
```
Comment (calebcartwright) [inline tests/source/hex_literal_lower.rs:1]:
```suggestion
// rustfmt-hex_literal_case: Lower
```

File: tests/source/hex_literal_upper.rs [hunk 7]
```diff
@@ -0,0 +1,5 @@
+// rustfmt-hex_literal_case: ToUpper
```
Comment (calebcartwright) [inline tests/source/hex_literal_upper.rs:1]:
```suggestion
// rustfmt-hex_literal_case: Upper
```

File: tests/target/hex_literal_ignore.rs [hunk 8]
```diff
@@ -0,0 +1,5 @@
+// rustfmt-hex_literal_case: Ignore
```
Comment (calebcartwright) [inline tests/target/hex_literal_ignore.rs:1]: Please also rename this file to account for the `ignore` --> `preserve` change
```suggestion
// rustfmt-hex_literal_case: Preserve
```

File: tests/source/hex_literal_upper.rs [hunk 9]
```diff
@@ -0,0 +1,5 @@
...
+fn main() {
+    let h1 = 0xCAFE_5EA7;
+    let h2 = 0xCAFE_F00Du32;
```
Comment (calebcartwright) [inline tests/source/hex_literal_upper.rs:4]: Since this is a case to highlight the letter casing being changed to upper I'd like a more lower-case centric input to make it easier on the eyes

e.g. 
```suggestion
    let h1 = 0xCaFE_5ea7;
    let h2 = 0xCAFE_F00Du32;
```

File: tests/target/hex_literal_lower.rs [hunk 10]
```diff
@@ -0,0 +1,5 @@
+// rustfmt-hex_literal_case: ToLower
```
Comment (calebcartwright) [inline tests/target/hex_literal_lower.rs:1]:
```suggestion
// rustfmt-hex_literal_case: Lower
```

File: tests/target/hex_literal_upper.rs [hunk 11]
```diff
@@ -0,0 +1,5 @@
+// rustfmt-hex_literal_case: ToUpper
```
Comment (calebcartwright) [inline tests/target/hex_literal_upper.rs:1]:
```suggestion
// rustfmt-hex_literal_case: Upper
```

**Review by calebcartwright — CHANGES_REQUESTED** [review]
This is in great shape, thanks so much! It does sound like a small feature at first, but very much looking forward to being able to get it released.

Will need a couple minor changes (largely renaming the variants and updating all the refs to them) but should be good to merge afterwards.

**Comment by calebcartwright** [comment]
Thanks for making these updates, though please note there's still at least one more that needs to be resolved before this can proceed

**Comment by calebcartwright** [comment]
Awesome thank you! Going to go ahead and merge this which will put it on track to be included the next time we update rustfmt. No  ETA as of yet, but there will be couple PRs to move across branches and ultimately update the submodule in the main compiler repo. GH should notify you on those given the authorship of the commits so you'll know once this hits nightly
